rpc: Add level 3 verbosity to getblock RPC call.

fyquah opened the pull request:
Display the prevout in transaction inputs when calling getblock level 3 verbosity. This PR affects the existing `/rest/block` API by adding a `prevout` fields to tx inputs. This is mentioned in the change to the release notes.

I added some functional tests that
- checks that the RPC call still works when TxUndo can't be found
- Doesn't display the "value" or "scriptPubKey" of the previous output when at a lower verbosity level

This "completes" the issue https://github.com/bitcoin/bitcoin/issues/18771

0xB10C commented:
Concept ACK. Thanks for picking this up! Will have a closer look and test this soon.

luke-jr commented on src/core_write.cpp:
```diff
@@ -223,6 +223,17 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
+            const CTxOut& prev_txout = txundo->vprevout[i].out;
+            UniValue p(UniValue::VOBJ);
+            UniValue o_script_pub_key(UniValue::VOBJ);
+
```
Where did "height" and "generated" go?

fyquah replied:
Added it back

kiminuo replied:
(Please mark as resolved 🙏 )

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -154,30 +154,39 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
+            const bool display_prev_out = (verbosity == BlockToJsonVerbosity::TX_DETAILS_AND_PREV_OUT);
             UniValue objTx(UniValue::VOBJ);
-            TxToUniv(*tx, uint256(), objTx, true, RPCSerializationFlags(), txundo);
+            std::cout << display_prev_out  << std::endl;
```
no thx

fyquah replied:
Whoop, sorry.

kiminuo replied:
(Please mark as resolved 🙏 )

luke-jr commented on src/core_io.h:
```diff
@@ -46,6 +46,6 @@ std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr, const bool display_prev_out = false);
```
Can we just use `BlockToJsonVerbosity` here? (Maybe rename it)

fyquah replied:
I renamed it to TxVerbosity, and moved it to src/core_io.h. Feels cleaner than passing bools.

kiminuo replied:
(Please mark as resolved 🙏 )

kiminuo replied:
(Please mark as resolved 🙏 )

luke-jr commented on src/core_write.cpp:
```diff
@@ -223,6 +223,28 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
+                UniValue o_script_pub_key(UniValue::VOBJ);
+
+                p.pushKV("value", ValueFromAmount(prev_txout.nValue));
+                p.pushKV("height", static_cast<uint64_t>(prev_coin.nHeight));
+                p.pushKV("generated", !static_cast<bool>(prev_coin.fCoinBase));
```
```suggestion
                p.pushKV("height", uint64_t(prev_coin.nHeight));
                p.pushKV("generated", bool(prev_coin.fCoinBase));
```

luke-jr commented on src/core_write.cpp:
```diff
@@ -223,6 +223,28 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
+                UniValue o_script_pub_key(UniValue::VOBJ);
+
+                p.pushKV("value", ValueFromAmount(prev_txout.nValue));
+                p.pushKV("height", static_cast<uint64_t>(prev_coin.nHeight));
+                p.pushKV("generated", !static_cast<bool>(prev_coin.fCoinBase));
```
Why did you invert this? That is wrong?

fyquah replied:
i was confused as to what "generated" means. Why don't we just call that field "coinbase" ?

luke-jr replied:
The coinbase is the *input* to the generation transaction (height + 95 bytes arbitrary). The *outputs* (which are being spent in the input here) are simply generated coins.

fyquah replied:
ah i see. this is fixed now, and i modified the tests to reflect this too.

kiminuo replied:
(Please mark as resolved 🙏 )

0xB10C commented on src/core_io.h:
```diff
@@ -20,6 +20,12 @@ class uint256;
 class UniValue;
 class CTxUndo;
 
+enum class TxVerbosity {
+    DONT_SHOW_TX_DETAILS,
+    TX_DETAILS,
+    TX_DETAILS_AND_PREV_OUT
+};
```
nit: I find something like this cleaner

```suggestion
enum class TxVerbosity {
    SHOW_TXID
    SHOW_DETAILS,
    SHOW_DETAILS_AND_PREV_OUT
};
```

fyquah replied:
done, adopted change.

Question with code review like this, is it preferred to make a separate commit, or squash everything to a single commit ?

0xB10C replied:
For small code review like the above 'Suggested change' through the GitHub web interface it's best to just fix it up locally, squash and then force push IMO. No reason to have it a separate commit (and I personally don't care about the co-author thing that GitHub does).

0xB10C reviewed (approved):
ACK 53ea5bd39bfb1da614c264656536edf954419373

Code reviewed and tested both the RPC and REST interface on testnet with a few blocks.

luke-jr commented:
Ideally, this should probably be 3 commits: one changing the bool to the class enum (with 2 values), one adding the new feature, and another adding tests.

fyquah committed:
rpc: Add test for level 3 verbosity getblock rpc call.

fyquah commented:
> Ideally, this should probably be 3 commits

Done

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -154,30 +154,36 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
-    if (txDetails) {
+
+    switch (verbosity) {
+      case TxVerbosity::SHOW_TXID:
```
indentation is wrong

kiminuo replied:
(Seems done; please mark as resolved 🙏 )

luke-jr commented on src/core_io.h:
```diff
@@ -46,6 +53,6 @@ std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr, TxVerbosity verbosity = TxVerbosity::SHOW_TXID);
```
Nit: This should be added in the later commit

luke-jr commented on src/rpc/blockchain.h:
```diff
@@ -40,7 +41,7 @@ double GetDifficulty(const CBlockIndex* blockindex);
 
 /** Block description to JSON */
-UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIndex* blockindex, bool txDetails = false) LOCKS_EXCLUDED(cs_main);
+UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIndex* blockindex, const TxVerbosity verbosity = TxVerbosity::SHOW_TXID) LOCKS_EXCLUDED(cs_main);
```
`const` doesn't do anything here, and is missing in the definition.

kiminuo replied:
(`const` was removed; please mark as resolved 🙏 )

luke-jr commented on src/core_io.h:
```diff
@@ -46,6 +53,6 @@ std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr, TxVerbosity verbosity = TxVerbosity::SHOW_TXID);
```
`SHOW_DETAILS` makes more sense as a default (`SHOW_TXID` isn't even supported by this function!)

kiminuo replied:
(Please mark as resolved 🙏 )

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -964,7 +971,17 @@ static RPCHelpMan getblock()
+    TxVerbosity block_to_json_verbosity;
+
+    if (verbosity <= 1) {
+      block_to_json_verbosity = TxVerbosity::SHOW_TXID;
```
indentation is wrong

kiminuo replied:
(Looks correct to me now; please mark as resolved 🙏 )

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -964,7 +971,17 @@ static RPCHelpMan getblock()
 
-    return blockToJSON(block, tip, pblockindex, verbosity >= 2);
+    TxVerbosity block_to_json_verbosity;
+
```
nit: I'd drop the blank link

luke-jr commented on src/core_write.cpp:
```diff
@@ -223,6 +223,28 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
+
+                case TxVerbosity::SHOW_DETAILS_AND_PREV_OUT:
+                    const Coin&   prev_coin  = txundo->vprevout[i];
+                    const CTxOut& prev_txout = prev_coin.out;
```
We already have `prev_txout` above. Move `prev_coin` up there?

fyquah replied:
Done

luke-jr replied:
This is still here...

kiminuo replied:
Is this still relevant?

luke-jr commented on src/rest.cpp:
```diff
@@ -295,12 +295,12 @@ static bool rest_block(HTTPRequest* req,
 static bool rest_block_extended(const util::Ref& context, HTTPRequest* req, const std::string& strURIPart)
 {
-    return rest_block(req, strURIPart, true);
+    return rest_block(req, strURIPart, TxVerbosity::SHOW_DETAILS_AND_PREV_OUT);
```
This could use a test

fyquah replied:
I added a test in 37f2c4e9e09d586e1c7d38332dc7fe7864245b97

fyquah committed:
rpc: Add level 3 verbosity to getblock RPC call.

Display the prevout in transaction inputs when calling getblock level 3
verbosity.

Co-authored-by: Luke Dashjr <luke_github1@dashjr.org>
Co-authored-by: 0xB10C <19157360+0xB10C@users.noreply.github.com>

fyquah committed:
rest: Add test for prevout fields in getblock

luke-jr commented on src/core_io.h:
```diff
@@ -46,6 +52,6 @@ std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr, TxVerbosity verbosity = TxVerbosity::SHOW_DETAILS);
```
This is one commit too early (causing a build failure in that commit)

fyquah replied:
Fixed. I moved the change in src/core_write.cpp from the previous commit to this one.

fyquah committed:
rpc: Replace boolean argument for tx details with enum class.

Co-authored-by: Luke Dashjr <luke_github1@dashjr.org>
Co-authored-by: 0xB10C <19157360+0xB10C@users.noreply.github.com>

luke-jr reviewed (changes requested):
Needs update for getblock docs

```
"If verbosity is 3, returns an Object with information about block <hash> and information about each transaction, including prevout information for inputs (only for blocks in the current best chain).\n",
```

fyquah commented:
> Needs update for getblock docs
> 
> ```
> "If verbosity is 3, returns an Object with information about block <hash> and information about each transaction, including prevout information for inputs (only for blocks in the current best chain).\n",
> ```

Done, I added to the commit that implements this feature

luke-jr reviewed (approved):
utACK bce09fe7429d8a007a3a96cd0cea28d5535f35cf

luke-jr commented:
(Could still be cleaner with the duplicate line removed https://github.com/bitcoin/bitcoin/pull/21245#discussion_r585016515)

luke-jr reviewed (approved):
utACK 7fc316e2c9f83cdab1b3b6cbd780c8d498f0bbb9

0xB10C commented:
ACK 7fc316e2c9f83cdab1b3b6cbd780c8d498f0bbb9.

- tested that both the RPC and REST interface have a new `prevout` field in the `vin`'s
- tested that the contents of the `getblock` RPC and `/rest/block/<hash>.json` are the same by piping them into `jq` and `diff`ing the results
- tested that a random `vin` can be found on chain as a previous output and the `height`, `value`, `address`, `type` and `scriptPubkey.hex` fields are correct with, for example, this bash helper:


``` bash
# make sure to play around with different block hashes and transaction indices (here 50) 
vin0=$(curl localhost:8332/rest/block/0000000000000000000c07a5ac516dc2a766da3bd1f85ec92090ba497188fd1b.json | jq .tx[50].vin[0])
vin0txid=$(echo $vin0 | jq .txid -r)
vin0nout=$(echo $vin0 | jq .vout)
vin0height=$(echo $vin0 | jq .prevout.height)
vin0blockhash=$(bitcoin-cli getblockhash $vin0height)

echo $vin0 | jq
echo ""
bitcoin-cli getrawtransaction $vin0txid true $vin0blockhash | jq .vout[$vin0nout]
```

----

**BlockToJsonVerbose** and **BlockToJsonVerboseWrite** (JSON serialization from #21170) benchmarks:

master @ cabe63759ce890a7d39d72f7b8046195b0edb421
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       71,551,280.00 |               13.98 |    0.1% |  561,473,856.00 |  136,734,366.00 |  4.106 | 105,739,596.00 |    0.4% |      0.79 | `BlockToJsonVerbose`
|       24,850,266.00 |               40.24 |    0.4% |  231,100,742.00 |   49,070,928.00 |  4.710 |  43,476,242.00 |    0.3% |      0.29 | `BlockToJsonVerboseWrite`



this PR @ 7fc316e2c9f83cdab1b3b6cbd780c8d498f0bbb9

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       72,740,207.00 |               13.75 |    0.5% |  561,708,569.00 |  138,848,957.00 |  4.045 | 105,777,166.00 |    0.4% |      0.80 | `BlockToJsonVerbose`
|       24,838,117.00 |               40.26 |    0.6% |  231,115,449.00 |   49,100,061.00 |  4.707 |  43,478,330.00 |    0.3% |      0.29 | `BlockToJsonVerboseWrite`

kiminuo reviewed (commented):
Some nitpicking from my side :)

kiminuo commented on test/functional/rpc_blockchain.py:
```diff
@@ -395,17 +395,55 @@ def _test_getblock(self):
         miniwallet.send_self_transfer(fee_rate=fee_per_kb, from_node=node)
         blockhash = node.generate(1)[0]
 
+        def check_fee_not_in_block(verbosity):
```
Maybe the function can be called `assert_fee_not_in_block`.

kiminuo commented on test/functional/rpc_blockchain.py:
```diff
@@ -395,17 +395,55 @@ def _test_getblock(self):
         self.log.info("Test that getblock with verbosity 1 doesn't include fee")
-        block = node.getblock(blockhash, 1)
-        assert 'fee' not in block['tx'][1]
+        check_fee_not_in_block(1)
```
It would be great to introduce a constant for this value (similar to this ones https://github.com/bitcoin/bitcoin/pull/21245/files#diff-0b117a50ea1a4d205e8b5268d01f06b100c6d75ccc743628774b1f656077a0c3R23). I think that they could be re-used and code would be self-describing.

fyquah replied:
i disagree -- i think what we are asserting here is that level 1 verbosity does not have fee in block,  level 2 & 3 does etc.. a constant for the value would probably make the test intention less clear.

kiminuo commented on test/functional/rpc_blockchain.py:
```diff
@@ -395,17 +395,55 @@ def _test_getblock(self):
+            assert 'fee' in tx
+            assert_equal(tx['fee'], tx['vsize'] * fee_per_byte)
+
+        def check_vin_contains_prevout(verbosity):
```
The parameter is always `3` so it seems unnecessary.

The function does more than checking that prevouts are present. Maybe `{verify|assert}_level3_block`?

fyquah replied:
I renamed the prefix to assert. The pattern i was going for was: `assert_<some-expected-behaviour>(verbosity_level)`, which is consistent with the other functions. I think i'd prefer leaving it as such.

kiminuo replied:
ok

theStack commented:
Concept ACK
Happy to review after rebase

kiminuo commented:
I have rebased the PR here: #22463. In my testing each individual commit can be compiled and `make check` passed for me. If you like the work, please just use it.

I plan to address review comments from this PR in #22463 in coming days, unless somebody beats me to it :)

fyquah commented:
sorry for the delay getting back to this. I just rebased -- it ought to be passing tests, i'll take a look at some of the reviews..

fyquah commented on src/core_write.cpp:
```diff
@@ -220,8 +220,27 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
+            const CTxOut& prev_txout = prev_coin.out;
+
             amt_total_in += prev_txout.nValue;
+            switch (verbosity) {
```
personally think it's clearer to comprehend, compared to something like `verbosity == TxVerbosity::SHOW_TXID || verbosity == TxVerbosity::SHOW_DETAILS)`

kiminuo replied:
But here it would be `if (verbosity == TxVerbosity::SHOW_DETAILS_AND_PREVOUT) { /* compute something */ }`. Seems very straightforward and Bitcoin Core does not use switches instead of simple if statements by default.

Obviously, this is rather a nit. So I don't want to waste your time with this. Feel free to resolve this suggestion if you are happy with the code as it is.

fyquah commented on src/rpc/blockchain.cpp:
```diff
@@ -964,7 +972,16 @@ static RPCHelpMan getblock()
+    TxVerbosity block_to_json_verbosity;
+    if (verbosity <= 1) {
+        block_to_json_verbosity = TxVerbosity::SHOW_TXID;
+    } else if (verbosity <= 2) {
```
Done

kiminuo replied:
Please mark this as resolved. 🙏 

(For some reason I can't do it and it's harder for me to track what is done, what is not done and what was rejected as a suggestion.)

fyquah commented on src/rpc/blockchain.cpp:
```diff
@@ -879,7 +886,8 @@ static RPCHelpMan getblock()
                 "\nIf verbosity is 0, returns a string that is serialized, hex-encoded data for block 'hash'.\n"
                 "If verbosity is 1, returns an Object with information about block <hash>.\n"
-                "If verbosity is 2, returns an Object with information about block <hash> and information about each transaction. \n",
+                "If verbosity is 2, returns an Object with information about block <hash> and information about each transaction. \n"
```
I removed the trailing space.

kiminuo replied:
(Please mark as resolved 🙏 )

fyquah commented on src/rpc/blockchain.h:
```diff
@@ -40,7 +41,7 @@ double GetDifficulty(const CBlockIndex* blockindex);
 
 /** Block description to JSON */
-UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIndex* blockindex, bool txDetails = false) LOCKS_EXCLUDED(cs_main);
+UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIndex* blockindex, TxVerbosity verbosity = TxVerbosity::SHOW_TXID) LOCKS_EXCLUDED(cs_main);
```
make sense, i removed it.

kiminuo replied:
(Please mark as resolved 🙏 )

fyquah commented:
@kiminuo  all your comments should been addressed now.

kiminuo commented:
> @kiminuo all your comments should been addressed now.

Thank you! There are still open review comments from my side. I'm not sure if you have seen them or not. Mostly nits though.

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -964,7 +972,16 @@ static RPCHelpMan getblock()
 
-    return blockToJSON(block, tip, pblockindex, verbosity >= 2);
+    TxVerbosity block_to_json_verbosity;
+    if (verbosity <= 1) {
```
(Please mark as resolved 🙏 )

kiminuo commented on src/core_write.cpp:
```diff
@@ -220,8 +220,27 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
         }
         if (calculate_fee) {
-            const CTxOut& prev_txout = txundo->vprevout[i].out;
+            const Coin&   prev_coin  = txundo->vprevout[i];
```
(Please mark as resolved 🙏 )

kiminuo commented on src/core_io.h:
```diff
@@ -20,6 +20,12 @@ class uint256;
+enum class TxVerbosity {
+    SHOW_TXID,
+    SHOW_DETAILS,
+    SHOW_DETAILS_AND_PREV_OUT
```
(Done; Please mark as resolved 🙏 )

kiminuo commented on src/core_write.cpp:
```diff
@@ -231,8 +231,27 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, bool include_add
+                    p.pushKV("value", ValueFromAmount(prev_txout.nValue));
+                    p.pushKV("height", uint64_t(prev_coin.nHeight));
+                    p.pushKV("generated", bool(prev_coin.fCoinBase));
+                    ScriptPubKeyToUniv(prev_txout.scriptPubKey, o_script_pub_key, true, true);
```
The convention seems to be:

```suggestion
                    ScriptPubKeyToUniv(prev_txout.scriptPubKey, o_script_pub_key, /* fIncludeHex */ true, /* include_addresses */ true);
```

fyquah committed:
release-notes: Add release note about getblock verbosity level 3.

fyquah commented on src/rest.cpp:
```diff
@@ -295,12 +295,12 @@ static bool rest_block(HTTPRequest* req,
 static bool rest_block_extended(const util::Ref& context, HTTPRequest* req, const std::string& strURIPart)
 {
-    return rest_block(req, strURIPart, true);
+    return rest_block(req, strURIPart, TxVerbosity::SHOW_DETAILS_AND_PREV_OUT);
```
good point. I added a commit with changes to the release notes, please have a look!

fyquah commented:
@kiminuo i think they should all be addressed now ..

kiminuo commented:
@fyquah Thanks! 

Is this Luke's comment addressed: https://github.com/bitcoin/bitcoin/pull/21245#discussion_r585016515 ? I'm not sure. 🤷‍♂️

kiminuo commented on doc/release-notes.md:
```diff
@@ -109,6 +109,15 @@ Updated RPCs
   of `decodescript` these fields are top-level attributes, and included again as attributes
   of the `scriptPubKey` object. (#20286)
 
+- The `getblock` rpc now supports a level 3 verbosity containing a transaction inputs'
```
Maybe:

```suggestion
- The `getblock` RPC command now supports verbose level 3 containing transaction inputs'
```
(naming based on https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.15.0.md)

fyquah commented:
> Is this Luke's comment addressed: #21245 (comment) ? I'm not sure. 🤷‍♂️

I don't think so. It has already been moved into the `if(calculate_fee)` block.

fyquah commented:
@kiminuo I think all but 2 remaining comments above are addressed.

fyquah commented:
Maintainers: I believe all the reviews are addressed. What else is blocking merge?

Once again, sorry for the delay getting back to this.

theStack reviewed (approved):
ACK cecbfed94e90df32f59ee661b83cb28f30c68587

Reviewed the code, built locally and ran the tests, issued a few `getblock` RPC commands on mainnet with verbose levels 1/2/3, LGTM. The tests could still be a bit more detailled, e.g. the sub-fields "height" and "scriptPubKey" are currently not covered (see also my comment beow), but this is non-blocking in my opinion.

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -199,30 +199,37 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
+            for (size_t i = 0; i < block.vtx.size(); ++i) {
+                const CTransactionRef& tx = block.vtx.at(i);
+                // coinbase transaction (i == 0) doesn't have undo data
+                const CTxUndo* txundo = (have_undo && i) ? &blockUndo.vtxundo.at(i - 1) : nullptr;
```
Nit:
```suggestion
                const CTxUndo* txundo = (have_undo && i > 0) ? &blockUndo.vtxundo.at(i - 1) : nullptr;
```

kiminuo reviewed (commented):
utACK

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -930,7 +937,8 @@ static RPCHelpMan getblock()
                 "If verbosity is 1, returns an Object with information about block <hash>.\n"
-                "If verbosity is 2, returns an Object with information about block <hash> and information about each transaction. \n",
+                "If verbosity is 2, returns an Object with information about block <hash> and information about each transaction.\n"
+                "If verbosity is 3, returns an Object with information about block <hash> and information about each transaction, including prevout information for inputs (only for blocks in the current best chain).\n",
```
FMI: "only for blocks in the current best chain" - Is this precisely worded with regard to how pruning works?

fyquah replied:
iiuc, txundo will not be looked up for pruned blocks. I reworded this from "only for blocks" to "only for unpruned blocks".

fyquah committed:
core_write: Rename calculate_fee to have_undo for clarity

fyquah commented:
@kiminuo i think your comments should be addressed now. Can you mark this as reviewed once this looks good?

kiminuo commented:
Currently `./bitcoin-cli -testnet -rpcuser=test -rpcpassword=test getblock` prints:


<details>
  <summary>Console output</summary>

```
error code: -1
error message:
getblock "blockhash" ( verbosity )

If verbosity is 0, returns a string that is serialized, hex-encoded data for block 'hash'.
If verbosity is 1, returns an Object with information about block <hash>.
If verbosity is 2, returns an Object with information about block <hash> and information about each transaction.
If verbosity is 3, returns an Object with information about block <hash> and information about each transaction, including prevout information for inputs (only for blocks in the current best chain).

Arguments:
1. blockhash    (string, required) The block hash
2. verbosity    (numeric, optional, default=1) 0 for hex-encoded data, 1 for a json object, and 2 for json object with transaction data

Result (for verbosity = 0):
"hex"    (string) A string that is serialized, hex-encoded data for block 'hash'

Result (for verbosity = 1):
{                                 (json object)
  "hash" : "hex",                 (string) the block hash (same as provided)
  "confirmations" : n,            (numeric) The number of confirmations, or -1 if the block is not on the main chain
  "size" : n,                     (numeric) The block size
  "strippedsize" : n,             (numeric) The block size excluding witness data
  "weight" : n,                   (numeric) The block weight as defined in BIP 141
  "height" : n,                   (numeric) The block height or index
  "version" : n,                  (numeric) The block version
  "versionHex" : "hex",           (string) The block version formatted in hexadecimal
  "merkleroot" : "hex",           (string) The merkle root
  "tx" : [                        (json array) The transaction ids
    "hex",                        (string) The transaction id
    ...
  ],
  "time" : xxx,                   (numeric) The block time expressed in UNIX epoch time
  "mediantime" : xxx,             (numeric) The median block time expressed in UNIX epoch time
  "nonce" : n,                    (numeric) The nonce
  "bits" : "hex",                 (string) The bits
  "difficulty" : n,               (numeric) The difficulty
  "chainwork" : "hex",            (string) Expected number of hashes required to produce the chain up to this block (in hex)
  "nTx" : n,                      (numeric) The number of transactions in the block
  "previousblockhash" : "hex",    (string, optional) The hash of the previous block (if available)
  "nextblockhash" : "hex"         (string, optional) The hash of the next block (if available)
}

Result (for verbosity = 2):
{                   (json object)
  ...,              Same output as verbosity = 1
  "tx" : [          (json array)
    {               (json object)
      ...,          The transactions in the format of the getrawtransaction RPC. Different from verbosity = 1 "tx" result
      "fee" : n     (numeric) The transaction fee in BTC, omitted if block undo data is not available
    },
    ...
  ]
}

Examples:
> bitcoin-cli getblock "00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getblock", "params": ["00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"]}' -H 'content-type: text/plain;' http://127.0.0.1:8332/
```
</details>


It would be great to:

* Add "Result (for verbosity = 3):" in the output.
* Modify "2. verbosity    (numeric, optional, default=1) 0 for hex-encoded data, 1 for a json object, and 2 for json object with transaction data" to include your new level 3.

kiminuo commented:
The following examples may be worth adding to your [original post](https://github.com/bitcoin/bitcoin/pull/21245#issue-576925964):

Examples of the output:

### Verbose level 0

```bash
./bitcoin-cli -testnet getblock 000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5 0
```

<details>
  <summary>Console output</summary>

  ```
  0400c020fbf5f342465b89fc746df587a07339e662a944aff4c286ca0b000000000000008387d05d584be6d61398c894427255227592c9e7c3c093759bafa4487e271532819c0b6100ec421969ccd91002010000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff5a03e67b1f04819c0b612f64656275672e636f6d2ffabe6d6d5647e44d08c48adec0ee855a3e78909f300bc4a4f9b79b01167ad2d107747c3501000000000000002c6c0d02af403792e06d14248417542f120093c2020000000000ffffffff04b92e9500000000001976a914190f8ddccb511a66f0cfc9a33204d9287e65fdf788ac0000000000000000266a24b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df0000000000000000266a24aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c100000000000000002b6a2952534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f757301200000000000000000000000000000000000000000000000000000000000000000f1087e960100000000010386fe743a8db99cd6db255f8505d9e20e67e03edbd411874816382415a053b1120100000000ffffffffa0785a72492fc87903eb4e07893fc769ef561db7020ba9b43e9c484619c637e10000000000ffffffff7dcd54a6c50865c34290454183ffa1a7f31aae3557045007a9075b022a4759020100000000ffffffff02408e2c000000000017a914d1c71b4df72064844eec341c86dc68cf81b9c2ec87eac5000000000000160014f2aeff3fd69c70bcff9c0709cc3ed1521868538002483045022100e45dd0be123e8edc98780e8fb1746cd5449d18399818052b66e8ae5fed26b0c502205e971594f3198009a436416c88f4aeea96beb07c713b71a60ffc043996da36e1012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e00248304502210083212c58e557e4b9b31bc9788bc3d08af280ea4b0376c00f54231f4fc6afd80e0220140601b13356014c924bca99ff0eba14ee200af496f9c9716bdd087023cbf1b001210396b46245502c3011f8e1b8b19aa9e8a40fff3153e225f619f57297914d93d4570247304402201d1bae75003e8986346c6df5aabe4ff820651cdc3c30038e3c2937c0393a70e30220528ac716e5e480b830e4b742c451b85cf4afbf53b20b982adf1b83423c79d634012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e000000000
  ```
</details>



### Verbose level 1

```bash
./bitcoin-cli -testnet getblock 000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5 1
```

<details>
  <summary>Console output</summary>

  ```json
  {
    "hash": "000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5",
    "confirmations": 10,
    "height": 2063334,
    "version": 549453828,
    "versionHex": "20c00004",
    "merkleroot": "3215277e48a4af9b7593c0c3e7c992752255724294c89813d6e64b585dd08783",
    "time": 1628150913,
    "mediantime": 1628149483,
    "nonce": 282709097,
    "bits": "1942ec00",
    "difficulty": 64177845.58487042,
    "chainwork": "000000000000000000000000000000000000000000000558fbc2a80220ea0e44",
    "nTx": 2,
    "previousblockhash": "000000000000000bca86c2f4af44a962e63973a087f56d74fc895b4642f3f5fb",
    "nextblockhash": "000000000000000119b44a12b31fb2e1d2baa9ad5bb7419428f62cc77a4685f5",
    "strippedsize": 598,
    "size": 959,
    "weight": 2753,
    "tx": [
      "f10ed21d238ce32aaa00384af709381e2ecf1e3914604b4f0fbd817b3477017c",
      "6fcf360bade1cae884cb0d36dfb92455b7cfdc037cfbab75f86dfbfe28f08b3e"
    ]
  }
  ```
</details>

### Verbose level 2

```bash
./bitcoin-cli -testnet getblock 000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5 2
```

<details>
  <summary>Console output</summary>

  ```json
  {
    "hash": "000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5",
    "confirmations": 10,
    "height": 2063334,
    "version": 549453828,
    "versionHex": "20c00004",
    "merkleroot": "3215277e48a4af9b7593c0c3e7c992752255724294c89813d6e64b585dd08783",
    "time": 1628150913,
    "mediantime": 1628149483,
    "nonce": 282709097,
    "bits": "1942ec00",
    "difficulty": 64177845.58487042,
    "chainwork": "000000000000000000000000000000000000000000000558fbc2a80220ea0e44",
    "nTx": 2,
    "previousblockhash": "000000000000000bca86c2f4af44a962e63973a087f56d74fc895b4642f3f5fb",
    "nextblockhash": "000000000000000119b44a12b31fb2e1d2baa9ad5bb7419428f62cc77a4685f5",
    "strippedsize": 598,
    "size": 959,
    "weight": 2753,
    "tx": [
      {
        "txid": "f10ed21d238ce32aaa00384af709381e2ecf1e3914604b4f0fbd817b3477017c",
        "hash": "9a614c99fc94459d6a5901b4091576365def3a23e002d77e6d79fd153d533bf4",
        "version": 1,
        "size": 357,
        "vsize": 330,
        "weight": 1320,
        "locktime": 2524842225,
        "vin": [
          {
            "coinbase": "03e67b1f04819c0b612f64656275672e636f6d2ffabe6d6d5647e44d08c48adec0ee855a3e78909f300bc4a4f9b79b01167ad2d107747c3501000000000000002c6c0d02af403792e06d14248417542f120093c2020000000000",
            "txinwitness": [
              "0000000000000000000000000000000000000000000000000000000000000000"
            ],
            "sequence": 4294967295
          }
        ],
        "vout": [
          {
            "value": 0.09776825,
            "n": 0,
            "scriptPubKey": {
              "asm": "OP_DUP OP_HASH160 190f8ddccb511a66f0cfc9a33204d9287e65fdf7 OP_EQUALVERIFY OP_CHECKSIG",
              "hex": "76a914190f8ddccb511a66f0cfc9a33204d9287e65fdf788ac",
              "address": "mhoTpa5QLRPFKsH4DdWSp1mru5BkEEWvYR",
              "type": "pubkeyhash"
            }
          },
          {
            "value": 0.00000000,
            "n": 1,
            "scriptPubKey": {
              "asm": "OP_RETURN b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df",
              "hex": "6a24b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df",
              "type": "nulldata"
            }
          },
          {
            "value": 0.00000000,
            "n": 2,
            "scriptPubKey": {
              "asm": "OP_RETURN aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c1",
              "hex": "6a24aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c1",
              "type": "nulldata"
            }
          },
          {
            "value": 0.00000000,
            "n": 3,
            "scriptPubKey": {
              "asm": "OP_RETURN 52534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f7573",
              "hex": "6a2952534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f7573",
              "type": "nulldata"
            }
          }
        ],
        "hex": "010000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff5a03e67b1f04819c0b612f64656275672e636f6d2ffabe6d6d5647e44d08c48adec0ee855a3e78909f300bc4a4f9b79b01167ad2d107747c3501000000000000002c6c0d02af403792e06d14248417542f120093c2020000000000ffffffff04b92e9500000000001976a914190f8ddccb511a66f0cfc9a33204d9287e65fdf788ac0000000000000000266a24b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df0000000000000000266a24aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c100000000000000002b6a2952534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f757301200000000000000000000000000000000000000000000000000000000000000000f1087e96"
      },
      {
        "txid": "6fcf360bade1cae884cb0d36dfb92455b7cfdc037cfbab75f86dfbfe28f08b3e",
        "hash": "891566037ba9d65bdc6bce5135d27d18607e7903979d8fa014d1f7118d618691",
        "version": 1,
        "size": 521,
        "vsize": 278,
        "weight": 1109,
        "locktime": 0,
        "vin": [
          {
            "txid": "12b153a015243816488711d4db3ee0670ee2d905855f25dbd69cb98d3a74fe86",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "3045022100e45dd0be123e8edc98780e8fb1746cd5449d18399818052b66e8ae5fed26b0c502205e971594f3198009a436416c88f4aeea96beb07c713b71a60ffc043996da36e101",
              "02b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e0"
            ],
            "sequence": 4294967295
          },
          {
            "txid": "e137c61946489c3eb4a90b02b71d56ef69c73f89074eeb0379c82f49725a78a0",
            "vout": 0,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "304502210083212c58e557e4b9b31bc9788bc3d08af280ea4b0376c00f54231f4fc6afd80e0220140601b13356014c924bca99ff0eba14ee200af496f9c9716bdd087023cbf1b001",
              "0396b46245502c3011f8e1b8b19aa9e8a40fff3153e225f619f57297914d93d457"
            ],
            "sequence": 4294967295
          },
          {
            "txid": "0259472a025b07a90750045735ae1af3a7a1ff8341459042c36508c5a654cd7d",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "304402201d1bae75003e8986346c6df5aabe4ff820651cdc3c30038e3c2937c0393a70e30220528ac716e5e480b830e4b742c451b85cf4afbf53b20b982adf1b83423c79d63401",
              "02b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e0"
            ],
            "sequence": 4294967295
          }
        ],
        "vout": [
          {
            "value": 0.02920000,
            "n": 0,
            "scriptPubKey": {
              "asm": "OP_HASH160 d1c71b4df72064844eec341c86dc68cf81b9c2ec OP_EQUAL",
              "hex": "a914d1c71b4df72064844eec341c86dc68cf81b9c2ec87",
              "address": "2NCNRjYqKz6j377wA328zjTqRGWBR59ygz1",
              "type": "scripthash"
            }
          },
          {
            "value": 0.00050666,
            "n": 1,
            "scriptPubKey": {
              "asm": "0 f2aeff3fd69c70bcff9c0709cc3ed15218685380",
              "hex": "0014f2aeff3fd69c70bcff9c0709cc3ed15218685380",
              "address": "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn",
              "type": "witness_v0_keyhash"
            }
          }
        ],
        "fee": 0.00011200,
        "hex": "0100000000010386fe743a8db99cd6db255f8505d9e20e67e03edbd411874816382415a053b1120100000000ffffffffa0785a72492fc87903eb4e07893fc769ef561db7020ba9b43e9c484619c637e10000000000ffffffff7dcd54a6c50865c34290454183ffa1a7f31aae3557045007a9075b022a4759020100000000ffffffff02408e2c000000000017a914d1c71b4df72064844eec341c86dc68cf81b9c2ec87eac5000000000000160014f2aeff3fd69c70bcff9c0709cc3ed1521868538002483045022100e45dd0be123e8edc98780e8fb1746cd5449d18399818052b66e8ae5fed26b0c502205e971594f3198009a436416c88f4aeea96beb07c713b71a60ffc043996da36e1012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e00248304502210083212c58e557e4b9b31bc9788bc3d08af280ea4b0376c00f54231f4fc6afd80e0220140601b13356014c924bca99ff0eba14ee200af496f9c9716bdd087023cbf1b001210396b46245502c3011f8e1b8b19aa9e8a40fff3153e225f619f57297914d93d4570247304402201d1bae75003e8986346c6df5aabe4ff820651cdc3c30038e3c2937c0393a70e30220528ac716e5e480b830e4b742c451b85cf4afbf53b20b982adf1b83423c79d634012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e000000000"
      }
    ]
  }
  ```
</details>



### Verbose level 3

```bash
./bitcoin-cli -testnet getblock 000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5 3
```

<details>
  <summary>Console output</summary>

  ```json
  {
    "hash": "000000000000001f682b188971cc1a121546be4e9d5baf22934fdc7f538288d5",
    "confirmations": 10,
    "height": 2063334,
    "version": 549453828,
    "versionHex": "20c00004",
    "merkleroot": "3215277e48a4af9b7593c0c3e7c992752255724294c89813d6e64b585dd08783",
    "time": 1628150913,
    "mediantime": 1628149483,
    "nonce": 282709097,
    "bits": "1942ec00",
    "difficulty": 64177845.58487042,
    "chainwork": "000000000000000000000000000000000000000000000558fbc2a80220ea0e44",
    "nTx": 2,
    "previousblockhash": "000000000000000bca86c2f4af44a962e63973a087f56d74fc895b4642f3f5fb",
    "nextblockhash": "000000000000000119b44a12b31fb2e1d2baa9ad5bb7419428f62cc77a4685f5",
    "strippedsize": 598,
    "size": 959,
    "weight": 2753,
    "tx": [
      {
        "txid": "f10ed21d238ce32aaa00384af709381e2ecf1e3914604b4f0fbd817b3477017c",
        "hash": "9a614c99fc94459d6a5901b4091576365def3a23e002d77e6d79fd153d533bf4",
        "version": 1,
        "size": 357,
        "vsize": 330,
        "weight": 1320,
        "locktime": 2524842225,
        "vin": [
          {
            "coinbase": "03e67b1f04819c0b612f64656275672e636f6d2ffabe6d6d5647e44d08c48adec0ee855a3e78909f300bc4a4f9b79b01167ad2d107747c3501000000000000002c6c0d02af403792e06d14248417542f120093c2020000000000",
            "txinwitness": [
              "0000000000000000000000000000000000000000000000000000000000000000"
            ],
            "sequence": 4294967295
          }
        ],
        "vout": [
          {
            "value": 0.09776825,
            "n": 0,
            "scriptPubKey": {
              "asm": "OP_DUP OP_HASH160 190f8ddccb511a66f0cfc9a33204d9287e65fdf7 OP_EQUALVERIFY OP_CHECKSIG",
              "hex": "76a914190f8ddccb511a66f0cfc9a33204d9287e65fdf788ac",
              "address": "mhoTpa5QLRPFKsH4DdWSp1mru5BkEEWvYR",
              "type": "pubkeyhash"
            }
          },
          {
            "value": 0.00000000,
            "n": 1,
            "scriptPubKey": {
              "asm": "OP_RETURN b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df",
              "hex": "6a24b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df",
              "type": "nulldata"
            }
          },
          {
            "value": 0.00000000,
            "n": 2,
            "scriptPubKey": {
              "asm": "OP_RETURN aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c1",
              "hex": "6a24aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c1",
              "type": "nulldata"
            }
          },
          {
            "value": 0.00000000,
            "n": 3,
            "scriptPubKey": {
              "asm": "OP_RETURN 52534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f7573",
              "hex": "6a2952534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f7573",
              "type": "nulldata"
            }
          }
        ],
        "hex": "010000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff5a03e67b1f04819c0b612f64656275672e636f6d2ffabe6d6d5647e44d08c48adec0ee855a3e78909f300bc4a4f9b79b01167ad2d107747c3501000000000000002c6c0d02af403792e06d14248417542f120093c2020000000000ffffffff04b92e9500000000001976a914190f8ddccb511a66f0cfc9a33204d9287e65fdf788ac0000000000000000266a24b9e11b6d0e9243317b288d087de11b8e6ef80f7c2103f4de492da071b36caa92112c85df0000000000000000266a24aa21a9ed863dde47b2d6f490300e63cb1db36a41a47052956ac4eb40e17a9d8e15d694c100000000000000002b6a2952534b424c4f434b3af72ef02a50ecbfc6f9e6e464c57881bc23da25fbf5254088f7e68366001f757301200000000000000000000000000000000000000000000000000000000000000000f1087e96"
      },
      {
        "txid": "6fcf360bade1cae884cb0d36dfb92455b7cfdc037cfbab75f86dfbfe28f08b3e",
        "hash": "891566037ba9d65bdc6bce5135d27d18607e7903979d8fa014d1f7118d618691",
        "version": 1,
        "size": 521,
        "vsize": 278,
        "weight": 1109,
        "locktime": 0,
        "vin": [
          {
            "txid": "12b153a015243816488711d4db3ee0670ee2d905855f25dbd69cb98d3a74fe86",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "3045022100e45dd0be123e8edc98780e8fb1746cd5449d18399818052b66e8ae5fed26b0c502205e971594f3198009a436416c88f4aeea96beb07c713b71a60ffc043996da36e101",
              "02b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e0"
            ],
            "prevout": {
              "generated": false,
              "height": 2063323,
              "value": 0.00544519,
              "scriptPubKey": {
                "asm": "0 f2aeff3fd69c70bcff9c0709cc3ed15218685380",
                "hex": "0014f2aeff3fd69c70bcff9c0709cc3ed15218685380",
                "address": "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn",
                "type": "witness_v0_keyhash",
                "addresses": [
                  "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn"
                ],
                "reqSigs": 1
              }
            },
            "sequence": 4294967295
          },
          {
            "txid": "e137c61946489c3eb4a90b02b71d56ef69c73f89074eeb0379c82f49725a78a0",
            "vout": 0,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "304502210083212c58e557e4b9b31bc9788bc3d08af280ea4b0376c00f54231f4fc6afd80e0220140601b13356014c924bca99ff0eba14ee200af496f9c9716bdd087023cbf1b001",
              "0396b46245502c3011f8e1b8b19aa9e8a40fff3153e225f619f57297914d93d457"
            ],
            "prevout": {
              "generated": false,
              "height": 2034020,
              "value": 0.01184228,
              "scriptPubKey": {
                "asm": "0 4aefc0691c23cddd7ac75d9ba3714e37c802088a",
                "hex": "00144aefc0691c23cddd7ac75d9ba3714e37c802088a",
                "address": "tb1qfthuq6guy0xa67k8tkd6xu2wxlyqyzy2ctc4fz",
                "type": "witness_v0_keyhash",
                "addresses": [
                  "tb1qfthuq6guy0xa67k8tkd6xu2wxlyqyzy2ctc4fz"
                ],
                "reqSigs": 1
              }
            },
            "sequence": 4294967295
          },
          {
            "txid": "0259472a025b07a90750045735ae1af3a7a1ff8341459042c36508c5a654cd7d",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "txinwitness": [
              "304402201d1bae75003e8986346c6df5aabe4ff820651cdc3c30038e3c2937c0393a70e30220528ac716e5e480b830e4b742c451b85cf4afbf53b20b982adf1b83423c79d63401",
              "02b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e0"
            ],
            "prevout": {
              "generated": false,
              "height": 2007043,
              "value": 0.01253119,
              "scriptPubKey": {
                "asm": "0 f2aeff3fd69c70bcff9c0709cc3ed15218685380",
                "hex": "0014f2aeff3fd69c70bcff9c0709cc3ed15218685380",
                "address": "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn",
                "type": "witness_v0_keyhash",
                "addresses": [
                  "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn"
                ],
                "reqSigs": 1
              }
            },
            "sequence": 4294967295
          }
        ],
        "vout": [
          {
            "value": 0.02920000,
            "n": 0,
            "scriptPubKey": {
              "asm": "OP_HASH160 d1c71b4df72064844eec341c86dc68cf81b9c2ec OP_EQUAL",
              "hex": "a914d1c71b4df72064844eec341c86dc68cf81b9c2ec87",
              "address": "2NCNRjYqKz6j377wA328zjTqRGWBR59ygz1",
              "type": "scripthash"
            }
          },
          {
            "value": 0.00050666,
            "n": 1,
            "scriptPubKey": {
              "asm": "0 f2aeff3fd69c70bcff9c0709cc3ed15218685380",
              "hex": "0014f2aeff3fd69c70bcff9c0709cc3ed15218685380",
              "address": "tb1q72h0707kn3cteluuquyuc0k32gvxs5uqawmvpn",
              "type": "witness_v0_keyhash"
            }
          }
        ],
        "fee": 0.00011200,
        "hex": "0100000000010386fe743a8db99cd6db255f8505d9e20e67e03edbd411874816382415a053b1120100000000ffffffffa0785a72492fc87903eb4e07893fc769ef561db7020ba9b43e9c484619c637e10000000000ffffffff7dcd54a6c50865c34290454183ffa1a7f31aae3557045007a9075b022a4759020100000000ffffffff02408e2c000000000017a914d1c71b4df72064844eec341c86dc68cf81b9c2ec87eac5000000000000160014f2aeff3fd69c70bcff9c0709cc3ed1521868538002483045022100e45dd0be123e8edc98780e8fb1746cd5449d18399818052b66e8ae5fed26b0c502205e971594f3198009a436416c88f4aeea96beb07c713b71a60ffc043996da36e1012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e00248304502210083212c58e557e4b9b31bc9788bc3d08af280ea4b0376c00f54231f4fc6afd80e0220140601b13356014c924bca99ff0eba14ee200af496f9c9716bdd087023cbf1b001210396b46245502c3011f8e1b8b19aa9e8a40fff3153e225f619f57297914d93d4570247304402201d1bae75003e8986346c6df5aabe4ff820651cdc3c30038e3c2937c0393a70e30220528ac716e5e480b830e4b742c451b85cf4afbf53b20b982adf1b83423c79d634012102b90ee04fcba937c82b5a45b7f0fff4fbf6f1f4de2d152a16db94d31b40f352e000000000"
      }
    ]
  }
  ```
</details>

kiminuo commented on test/functional/interface_rest.py:
```diff
@@ -312,6 +312,16 @@ def run_test(self):
                             if 'coinbase' not in tx['vin'][0]}
         assert_equal(non_coinbase_txs, set(txs))
 
+        # Verify that the non-coinbase tx has "prevout" field set
```
nit:
```suggestion
        # Verify that the non-coinbase tx has "prevout" key set
```
A Python dictionary is a set of key-value pairs.

kiminuo commented on test/functional/interface_rest.py:
```diff
@@ -312,6 +312,16 @@ def run_test(self):
+            for vin in tx_obj["vin"]:
+                if "coinbase" not in vin:
+                    assert "prevout" in vin
+                    prevout = vin["prevout"]
+                    assert_equal(prevout["generated"], False)
```
Given that `prevout` variable is used just once, it may be shorter and actually easier to read:

```suggestion
                    assert_equal(vin["prevout"]["generated"], False)
```

kiminuo commented on test/functional/rpc_blockchain.py:
```diff
@@ -412,17 +412,55 @@ def _test_getblock(self):
+        def assert_vin_contains_prevout(verbosity):
+            block = node.getblock(blockhash, verbosity)
+            tx = block["tx"][1]
+            total_vin = 0
```
I believe this should be more like:

```suggestion
            total_vin = Decimal("0.00000000")
```

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -199,30 +199,37 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
+
+            for (size_t i = 0; i < block.vtx.size(); ++i) {
+                const CTransactionRef& tx = block.vtx.at(i);
+                // coinbase transaction (i == 0) doesn't have undo data
```
nit:
```suggestion
                // coinbase transaction (i.e. i == 0) doesn't have undo data
```

kiminuo commented on src/core_write.cpp:
```diff
@@ -230,9 +230,28 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, bool include_add
+
+                case TxVerbosity::SHOW_DETAILS_AND_PREVOUT:
+                    UniValue o_script_pub_key(UniValue::VOBJ);
+                    ScriptPubKeyToUniv(prev_txout.scriptPubKey, o_script_pub_key, /* includeHex */ true, /* include_addresses */ true);
```
Question: There is `include_addresses` parameter which is not used here. `true` is passed instead. Is that correct or should it look like:

```suggestion
                    ScriptPubKeyToUniv(prev_txout.scriptPubKey, o_script_pub_key, /* includeHex */ true, include_addresses);
```

It looks to me that `include_addresses = true` is deprecated actually.

luke-jr replied:
Since this is adding new output, I think it should probably just be a constant `false` here, actually.

kiminuo reviewed (commented):
Did some more review, tested it and the PR is shaping up very nicely.

It would be great if other reviewers could have a look too (@jonatack, @theStack possibly?) as I believe this has already received significant amount of work.

kiminuo commented:
@fyquah I have attempted to address my comments above and add RPC help here: https://github.com/kiminuo/bitcoin/tree/feature/2021-08-08-verbosity-level-3-getblock-iteration-3:

```bash
./bitcoin-cli -testnet getblock
```

reports for me:

<details>
  <summary>Console output</summary>

```
error code: -1
error message:
getblock "blockhash" ( verbosity )

If verbosity is 0, returns a string that is serialized, hex-encoded data for block 'hash'.
If verbosity is 1, returns an Object with information about block <hash>.
If verbosity is 2, returns an Object with information about block <hash> and information about each transaction.
If verbosity is 3, returns an Object with information about block <hash> and information about each transaction, including prevout information for inputs (only for unpruned blocks in the current best chain).

Arguments:
1. blockhash    (string, required) The block hash
2. verbosity    (numeric, optional, default=1) 0 for hex-encoded data, 1 for a json object, and 2 for json object with transaction data

Result (for verbosity = 0):
"hex"    (string) A string that is serialized, hex-encoded data for block 'hash'

Result (for verbosity = 1):
{                                 (json object)
  "hash" : "hex",                 (string) the block hash (same as provided)
  "confirmations" : n,            (numeric) The number of confirmations, or -1 if the block is not on the main chain
  "size" : n,                     (numeric) The block size
  "strippedsize" : n,             (numeric) The block size excluding witness data
  "weight" : n,                   (numeric) The block weight as defined in BIP 141
  "height" : n,                   (numeric) The block height or index
  "version" : n,                  (numeric) The block version
  "versionHex" : "hex",           (string) The block version formatted in hexadecimal
  "merkleroot" : "hex",           (string) The merkle root
  "tx" : [                        (json array) The transaction ids
    "hex",                        (string) The transaction id
    ...
  ],
  "time" : xxx,                   (numeric) The block time expressed in UNIX epoch time
  "mediantime" : xxx,             (numeric) The median block time expressed in UNIX epoch time
  "nonce" : n,                    (numeric) The nonce
  "bits" : "hex",                 (string) The bits
  "difficulty" : n,               (numeric) The difficulty
  "chainwork" : "hex",            (string) Expected number of hashes required to produce the chain up to this block (in hex)
  "nTx" : n,                      (numeric) The number of transactions in the block
  "previousblockhash" : "hex",    (string, optional) The hash of the previous block (if available)
  "nextblockhash" : "hex"         (string, optional) The hash of the next block (if available)
}

Result (for verbosity = 2):
{                   (json object)
  ...,              Same output as verbosity = 1
  "tx" : [          (json array)
    {               (json object)
      ...,          The transactions in the format of the getrawtransaction RPC. Different from verbosity = 1 "tx" result
      "fee" : n     (numeric) The transaction fee in BTC, omitted if block undo data is not available
    },
    ...
  ]
}

Result (for verbosity = 3):
{                                        (json object)
  ...,                                   Same output as verbosity = 2
  "tx" : [                               (json array)
    {                                    (json object)
      "vin" : [                          (json array)
        {                                (json object)
          ...,                           The same output as verbosity = 2
          "prevout" : {                  (json object)
            "generated" : true|false,    (boolean) Coinbase or not
            "height" : n,                (numeric) The height of the prevout
            "value" : n,                 (numeric) The value in BTC
            "scriptPubKey" : {           (json object)
              "asm" : "str",             (string) The asm
              "hex" : "str",             (string) The hex
              "reqSigs" : n,             (numeric, optional) (DEPRECATED, returned only if config option -deprecatedrpc=addresses is passed) Number of required signatures
              "type" : "str",            (string) The type, eg 'pubkeyhash'
              "address" : "str",         (string, optional) bitcoin address (only if a well-defined address exists)
              "addresses" : [            (json array, optional) (DEPRECATED, returned only if config option -deprecatedrpc=addresses is passed) Array of bitcoin addresses
                "str",                   (string) bitcoin address
                ...
              ]
            }
          }
        },
        ...
      ]
    },
    ...
  ]
}

Examples:
> bitcoin-cli getblock "00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getblock", "params": ["00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"]}' -H 'content-type: text/plain;' http://127.0.0.1:8332/
```
</details>

If you find the modifications (even partially) useful, please include it in your PR.

theStack reviewed (commented):
re-ACK 72dbe9816482d46c63168109c8745d3ec615b031

Verified via `git range-diff cecbfed9...72dbe981` that since my last ACK all requested changes since then (e.g. rename calculate_fee to have_undo (https://github.com/bitcoin/bitcoin/pull/21245#discussion_r672221707), more detailled "prevout" checking w.r.t. to coinbase/non-coinbase txs (https://github.com/bitcoin/bitcoin/pull/21245#discussion_r672184180), RPC help improvement (https://github.com/bitcoin/bitcoin/pull/21245#discussion_r672238121)) were tackled. Also compiled locally and successfully ran the involved functional tests.

> Did some more review, tested it and the PR is shaping up very nicely.

Agree! :)

@kiminuo made some [very good points for further improvements](https://github.com/bitcoin/bitcoin/pull/21245#pullrequestreview-724033069), especially considering the missing result desription for the newly introduced level in the RPC help. Since none of them are really blockers IMHO and the PR author does seem to be quite busy, maybe they can be done in a follow-up PR, to move things forward?

fyquah commented:
Hi, sorry, I haven't been able to spend much time on this :(

I like the suggestion to have @kiminuo 's follow-up PR addressing the non-blocking issues. Is there anything that is blocking merge at this point?

(Tagging an arbitrary maintainer: @fanquake )

fanquake commented:
> Is there anything that is blocking merge at this point?

This has only been ACK'd by a single reviewer.

luke-jr commented:
I would re-ACK except for https://github.com/bitcoin/bitcoin/pull/21245#discussion_r684024088

kiminuo commented:
@fyquah I think https://github.com/bitcoin/bitcoin/pull/21245#discussion_r684024088 is important so I would ACK once this is fixed. 

Also, the work is actually done here https://github.com/bitcoin/bitcoin/pull/21245#issuecomment-894853784 I believe (the last commit is extra and is for a follow-up PR).

kiminuo commented:
@fyquah Could you find a moment to address https://github.com/bitcoin/bitcoin/pull/21245#discussion_r684024088?

(PS: I'm willing to take over your PR to work on it if you are too busy. Anyway, it seems to me one little change is needed here, nothing time consuming.)

jaysonmald35 commented on src/rpc/blockchain.cpp:
```diff
@@ -1017,7 +1025,16 @@ static RPCHelpMan getblock()
     }
 
-    return blockToJSON(block, tip, pblockindex, verbosity >= 2);
+    TxVerbosity block_to_json_verbosity;
```
Thank u

fanquake commented:
Closing this for now that #22918 is open.
